pkg/backup/resource_backupper.go

```diff
@@ -237,43 +239,86 @@ func (rb *defaultResourceBackupper) backupResource(group *metav1.APIResourceList
 
 		log.Infof("Retrieved %d items", len(items))
 
+		// back up CRD for resource if found. We should only need to do this is IncludeClusterResources is nil. If it's false
+		// we don't want to back it up, and if it's true it will already be included.
+		if len(items) > 0 && rb.backupRequest.Spec.IncludeClusterResources == nil {
```

Comment:
Another thing to think about: even if `len(items) > 0`, we're not necessarily actually backing up any of those items, as the item backupper will check all the filters, including the `exclude-from-backup` label. So, it's possible that we end up not actually backing up any instances of the CRD, but then still backing up the CRD itself.

I think ideally, we only back up the CRD if we *actually* back up 1+ instances of it. One way we could accomplish this is to change the `backupItem` method to also return a boolean indicating whether the item was actually backed up or not. Then here in the resource backupper, we keep track of whether any call to `backupItem` returned a `true` there, and if so, we back up the CRD.

Reply:
I guess there are actually two ways of considering this. Either we only copy the CRD if at least one of its CRs is actually copied (which is what you suggest here), or we always copy the CRD, even if there are no instances of it here (possibly excluding it if the associated resource is explicitly excluded from backup). My concern with not copying the CRD is if you're backing up and restoring an application with an associated CRD, wouldn't you want the CRD included even if you don't happen to have any CRs at the moment. What if, for example, you have a CRD that uses CRs for a form of work queue. The application grabs a CR, does something, and then deletes the CR. If the queue is empty when you do the backup, your restore won't include the CRD. Then again, if we did this I'm not sure how "include CRDs even if there are no CRs" would differ from CRD behavior if we set includeClusterResources to 'true'. EIther way, `len(items) > 0` is the wrong test, since, as you point out, it doesn't take into account excluded resources.

Reply:
> wouldn't you want the CRD included even if you don't happen to have any CRs at the moment. What if, for example, you have a CRD that uses CRs for a form of work queue.

This is a great point actually, though if there are no CRs at the moment, we won't be able to derive the CRD to backup here anyway, right? I think in these cases, the CRD will need to be explicitly included (either with `--include-resources` or `--include-cluster-resources`).